Initial Debian Packaging

balbirthomas opened the pull request:
This source package builds both the libsmu shared library and its
python bindings. The installed python bindings and shared library
have been tested and alice-1.1 seems to work with them.

A few changes needed to be made to upstream sources for the package to build successfully on Debian/Jessie. However these changes were made using quilt and patches placed in the debian/patches directory so that upstream sources remain pristine. If agreeable some of these patches can be pushed upstream.

balbirthomas committed:
Initial Debian Packaging

This source package builds both the libsmu shared library and its
python bindings. The installed python bindings and shared library
have been tested and alice-1.1 seems to work with them.

balbirthomas committed:
Install udev rules in Debian package

This commit ensures that the udev rules for Analog Devices ADALM1000 is
installed into the libsmu0 shared library package. Hence any application
trying to utilise the shared library will be able connect to the device.

radhermit commented on debian/patches/0003-force-cpp11.patch:
```diff
@@ -0,0 +1,11 @@
+ 
+ if(CMAKE_COMPILER_IS_GNUCXX OR CMAKE_CXX_COMPILER_ID MATCHES "Clang")
+-	set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -pedantic")
++	set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -pedantic -std=c++11")
```
This is already enabled by set(CMAKE_CXX_STANDARD 11); however, if we really must lower the minimum required version to accommodate Debian stable that should be removed and probably make the flag conditionally used for the compilers that recognize it (i.e. it used to be set this way until I got annoyed with the warnings thrown when compiling under VS on Windows).

balbirthomas replied:
Good point. I have just pushed a new commit that makes using "-std=c++11" explicitly conditional on CMAKE version less than 3.1 and only when the compiler is GCC.

radhermit commented on debian/patches/0004-remove-absolute-lib-include-paths.patch:
```diff
@@ -0,0 +1,12 @@
+-	PUBLIC_HEADER DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}")
++	ARCHIVE DESTINATION "lib"
++	LIBRARY DESTINATION "lib"
++	PUBLIC_HEADER DESTINATION "include")
```
Shouldn't this already be handled by the GNUInstallDirs include? At least for recent versions of cmake, GNUInstallDirs special cases debian multiarch handling. You'd have to check out 3.0 or whatever version to see if it made it into that version as well.

I'd prefer using this method along with GNUInstallDirs since this change isn't entirely compatible with methodologies used on other distros.

balbirthomas replied:
The problem is that ```${CMAKE_INSTALL_INCLUDEDIR}``` resolves to a full path. So if the Debian package is being built in ```/foo/bar``` it resolves to something like ```/foo/bar/debian/tmp/foo/bar/usr``` etc. Ideally it should resolve to ```debian/tmp/usr``` and so on.  This causes the Debian packaging scripts (dh_install) to fail as it looks for files under ```debian/tmp/```.  This can remain a Debian only patch in the ```debian/patches``` directory if so desired. I wonder if there are more compatible ways to implement the CMake scripts so it builds on all distros, only using ```${CMAKE_INSTALL_PREFIX}``` to customise the build paths. Unfortunately I may not have the time to investigate this further. There is an additional issue mentioned in the ```debian/TODO``` file, that is spaces in directory names break the ```install``` target. Again I am not sure when I would be able to look into this.

radhermit commented on debian/patches/0001-fix-quote-cxx-flags.patch:
```diff
@@ -0,0 +1,11 @@
++	set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -pedantic")
+ endif()
+ set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -DDEBUG")
+ 
```
Already fixed a month or two ago in a rework branch, I'll backport the fix.

balbirthomas replied:
No problem. You can always edit ```debian/patches/series``` to remove this patch from the list and perhaps additionally delete ```0001-fix-quote-cxx-flags.patch```, if and when you choose to backport the fix.

radhermit replied:
I backported the fix right after I wrote that. :smile:

balbirthomas replied:
This may cause a merge conflict with this pull request.

radhermit commented:
Thanks for the initial work here! Just to note, there is a bunch of pending work happening towards restructuring the underlying architecture so certain build details will change when things are merged.

balbirthomas committed:
Make CXX standard setting method conditional on CMake version

This commit makes the method of enfocing GCC to use the C++11 standard
condition on the version of CMAKE. Older version of CMake (such as 3.0)
do not support the CXX_STANDARD options.

balbirthomas committed:
Remove fix CMake quoting patch and change calib location

This comit rebases to current master. In doing so it removes the
now unnecessary patch to fix CMake quoting of CXX flags. It also
changes the location from where calib.txt is taken.

radhermit commented on debian/patches/0003-use-explicit-cxx-std-on-older-cmake.patch:
```diff
@@ -0,0 +1,17 @@
+ # C++11 is required
+-set(CMAKE_CXX_STANDARD 11)
+-set(CXX_STANDARD_REQUIRED ON)
++if(CMAKE_COMPILER_IS_GNUCXX AND (CMAKE_VERSION VERSION_LESS 3.1))
```
I think this should probably accept clang as well as gcc.

balbirthomas replied:
Feel free to change it as you please. I may not be able to spend much more time with this but will appreciate it if you can move it along. Alternatively if you like me to close this pull request, i am ok with that too.

Commit:
Remove obsolete Debian patches

Upstream made changes that makes the some of the Debian patches
no longer necessary.
